### PR TITLE
Dynamic UI

### PR DESCRIPTION
**Pull request by Macroz**
Example of how to move actions to their own file by implementing a skeleton comment request.

### CODE REVIEW

File: src/cljs/rems/actions/action.cljs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+(defn- action-comment [label-title comment on-comment]
+  [:div.form-group
+   [:label {:for "judge-comment"} label-title]
+   [textarea {:id "judge-comment"
+              :name "judge-comment" :placeholder (text :t.actions/comment-placeholder)
+              :value comment
+              :on-change on-comment}]])
+
+(defn action-form-view [id title comment-title button content comment on-comment]
```
Comment (opqdonut) [inline src/cljs/rems/actions/action.cljs]: this could use a docstring

File: src/cljs/rems/actions/request_comment.cljs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+ (fn [db [_ value]] (assoc db ::comment value)))
+
+(defn- send-request-comment! [commenters user application-id comment description]
+  (post! "/api/applications/request_comment"
+         {:params {:application-id application-id
+                   :comment comment
+                   :recipients (map :userid commenters)}
+          :handler (fn [resp]
+                     ;; TODO use callbacks so no dependency?
+                     (rf/dispatch [:rems.application/set-status {:status :saved
```
Comment (opqdonut) [inline src/cljs/rems/actions/request_comment.cljs]: could trigger some sort of refresh-application event here, right?

File: src/cljs/rems/spa.cljs [hunk 3]
```diff
@@ -79,8 +79,7 @@
 (reg-sub
  :roles
  (fn [db _]
-   (when-let [roles (get-in db [:identity :roles])]
-     (set roles))))
```
Comment (opqdonut) [inline src/cljs/rems/spa.cljs]: was this change intentional?

Comment (Macroz) [reply src/cljs/rems/spa.cljs]: Yes, see the commit. It's better to already originally store the set and not just use subscription because many places use the `:roles` from db directly.

File: src/cljs/rems/actions/action.cljs (hunk 1 above)
Comment (Macroz) [reply src/cljs/rems/actions/action.cljs]: Yes definitely all the common stuff needs to be documented. 

Probably this can be refactored to be simpler with less params or keyword params.

File: src/cljs/rems/spa.cljs [hunk 4]
```diff
@@ -329,7 +328,7 @@
   [user-and-roles]
   (let [user-and-roles (js->clj user-and-roles :keywordize-keys true)]
     (rf/dispatch-sync [:set-identity (if (:user user-and-roles)
-                                       (update user-and-roles :roles #(mapv keyword (:roles user-and-roles)))
+                                       (update user-and-roles :roles (comp set #(mapv keyword (:roles user-and-roles))))
```
Comment (opqdonut) [inline src/cljs/rems/spa.cljs]: this is really weird code, why not just
```
(assoc user-and-roles :roles (set (map keyword (:roles user-and-roles))))

Comment (Macroz) [reply src/cljs/rems/spa.cljs]: I don't know, it's somebody's old code but can be simplified.

File: src/cljs/rems/actions/request_comment.cljs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+
+;; TODO together with application.cljs
+(defn enrich-user [user]
+  (assoc user :display (str (:name user) " (" (:email user) ")")))
+
+(rf/reg-event-db
+ ::set-potential-commenters
+ (fn [db [_ users]]
+   (assoc db ::potential-commenters (map enrich-user users))))
```
Comment (luontola) [inline src/cljs/rems/actions/request_comment.cljs]: `::potential-commenters` is initialized to be a set, but this makes it a seq.

File: src/cljs/rems/actions/request_comment.cljs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+ ::set-selected-commenters
+ (fn [db [_ users]]
+   (assoc db ::selected-commenters users)))
+
+(rf/reg-event-db
+ ::add-selected-commenter
+ (fn [db [_ user]]
+   (if (contains? (::selected-commenters db) user)
+     db
+     (update db ::selected-commenters conj user))))
```
Comment (luontola) [inline src/cljs/rems/actions/request_comment.cljs]: We could avoid this `if` if `::selected-commenters` were be a set.

File: src/cljs/rems/actions/action.cljs (hunk 1 above)
Comment (luontola) [reply src/cljs/rems/actions/action.cljs]: +1 for keyword params when there are over 3 params.

File: src/cljs/rems/actions/request_comment.cljs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+(rf/reg-sub ::selected-commenters (fn [db _] (::selected-commenters db)))
+(rf/reg-sub ::comment (fn [db _] (::comment db)))
+
+(rf/reg-event-db
+ ::set-comment
+ (fn [db [_ value]] (assoc db ::comment value)))
+
+(defn- send-request-comment! [commenters user application-id comment description]
+  (post! "/api/applications/request_comment"
```
Comment (luontola) [inline src/cljs/rems/actions/request_comment.cljs]: Should this use the command api?

File: src/cljs/rems/actions/request_comment.cljs (hunk 2 above)
Comment (luontola) [reply src/cljs/rems/actions/request_comment.cljs]: Do you mean a refresh-application which would replace the enter-application-page event it dispatches now?

Comment (Macroz) [reply src/cljs/rems/actions/request_comment.cljs]: I think the use site in `application.cljs` should provide the refresh dispatching function to this component in a `on-success` callback. 

Set status can be refactored to each form. It does not need to stay in the `application.cljs`.

File: src/cljs/rems/actions/request_comment.cljs (hunk 5 above)
Comment (Macroz) [reply src/cljs/rems/actions/request_comment.cljs]: Some kind of old bug

File: src/cljs/rems/actions/request_comment.cljs (hunk 6 above)
Comment (Macroz) [reply src/cljs/rems/actions/request_comment.cljs]: Yes it makes sense, it should be a set.

File: src/cljs/rems/actions/request_comment.cljs (hunk 7 above)
Comment (Macroz) [reply src/cljs/rems/actions/request_comment.cljs]: Yes it's already done in my latest version, this WIP was just for comment for the general direction.

File: src/cljs/rems/actions/action.cljs (hunk 1 above)
Comment (Macroz) [reply src/cljs/rems/actions/action.cljs]: Done in latest.

File: src/clj/rems/api/applications.clj [hunk 8]
```diff
@@ -187,6 +208,12 @@
       :return Reviewers
       (ok (get-reviewers)))
 
+    (GET "/commenters" []
+      :summary "Available third party commenters"
```
[the remark this answers is not in the document]
Comment (Macroz) [reply src/clj/rems/api/applications.clj]: I think it's just waste to add the third party everwhere.

File: src/clj/rems/api/applications.clj [hunk 9]
```diff
@@ -136,18 +144,31 @@
...
 
+;; TODO Filter applicant, requesting user
 (defn get-reviewers []
   (for [u (->> (users/get-all-users)
                (remove invalid-reviewer?))]
     {:userid (get u "eppn")
      :name (get u "commonName")
      :email (get u "mail")}))
 
+(defn invalid-commenter? [u]
```
[the remark this answers is not in the document]
Comment (Macroz) [reply src/clj/rems/api/applications.clj]: Well it's an exact duplicate of the invalid-reviewer? so someone can refactor them later.

[the remark this answers is not in the document]
Comment (Macroz) [reply src/clj/rems/api/applications.clj]: There's now a TODO about the refactoring to remove duplication
